packages/strapi-parts/src/NumberInput/NumberInput.js

```diff
@@ -0,0 +1,155 @@
+import React, { useEffect, useState, useRef } from 'react';
+import PropTypes from 'prop-types';
+import FilterDropdownIcon from '@strapi/icons/FilterDropdownIcon';
+import styled from 'styled-components';
+import { NumberFormatter, NumberParser } from '@internationalized/number';
+import { Field, FieldLabel, FieldHint, FieldError, FieldInput } from '../Field';
+import { Stack } from '../Stack';
+import { Row } from '../Row';
+import { Box } from '../Box';
+import { useId } from '../helpers/useId';
+import { KeyboardKeys } from '../helpers/keyboardKeys';
+import { getDefaultLocale } from '../helpers/getDefaultLocale';
+
+const ArrowButton = styled.button`
+  display: flex;
+  height: 1rem;
+  align-items: ${({ reverse }) => (reverse ? 'flex-end' : 'flex-start')};
+  margin-bottom: ${({ reverse, theme }) => (reverse ? theme.spaces[1] : 0)};
+
+  svg {
+    display: block;
+    height: ${4 / 16}rem;
+    transform: ${({ reverse }) => (reverse ? 'rotateX(180deg)' : undefined)};
+  }
+`;
+
+export const NumberInput = React.forwardRef(
+  ({ startAction, name, hint, error, label, labelAction, id, onValueChange, value, step, ...props }, ref) => {
+    const [inputValue, setInputValue] = useState(value || '');
+    const generatedId = useId('numberinput', id);
+    const numberParserRef = useRef(new NumberParser(getDefaultLocale()));
+    const numberFormaterRef = useRef(new NumberFormatter(getDefaultLocale()));
+
+    useEffect(() => {
+      if (value !== undefined) {
+        setInputValue(numberFormaterRef.current.format(value));
+      }
+    }, [value]);
+
+    const handleChange = (e) => {
+      if (numberParserRef.current.isValidPartialNumber(e.target.value)) {
+        setInputValue(e.target.value);
+      }
+    };
```

Comment:
ℹ️ We only modify the internal value on change. The external one will be modified on blur and then formatted back thanks to the useEffect call